File: 496.Next-Greater-Element-I.go
// https://leetcode.com/problems/next-greater-element-i/description/
//
// algorithms
// Easy (57.62%)
// Total Accepted:    75.6k
// Total Submissions: 131.2k

package leetcode

func nextGreaterElement(findNums []int, nums []int) []int {
	res := make([]int, 0, len(findNums))
	dict := make(map[int]int)
	for k, v := range nums {
		dict[v] = k
	}

	for i := 0; i < len(findNums); i++ {
		j := dict[findNums[i]] + 1
		for {
			if j == len(nums) {
				res = append(res, -1)
				break
			}
			if nums[j] > findNums[i] {
				res = append(res, nums[j])
				break
			}
			j++
		}
	}
	return res
}

File: 496.Next-Greater-Element-I.java
// https://leetcode.com/problems/next-greater-element-i/
//
// algorithms
// Easy (61.91%)
// Total Accepted: 127,813
// Total Submissions: 206,445


class Solution {

    private static final int PLACEHOLDER = -1;

    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        HashMap<Integer, Integer> map = new HashMap<>();
        LinkedList<Integer> stack = new LinkedList<>();

        for (int n : nums2) {
            while (!stack.isEmpty() && n > stack.peekLast()) {
                map.put(stack.pollLast(), n);
            }

            stack.offer(n);
        }

        int[] res = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            if (map.containsKey(nums1[i])) {
                res[i] = map.get(nums1[i]);
            } else {
                res[i] = PLACEHOLDER;
            }
        }

        return res;
    }
}